lib/src/dictionary.dart

```diff
@@ -0,0 +1,524 @@
+      final removedNode = node.remove(shift + 5, hash, key);
+      final newNodes = new List<Node<K, V>>.from(_nodes, growable: false);
+      newNodes[index] = removedNode;
+      return new ArrayNode(newNodes, _shift, _size - 1);
+    } else {
+      return this;
+    }
+  }
+
+  @override
+  Iterable<_Tuple2<K, V>> traverseBoth() sync* {
+    for (final node in _nodes) {
+      yield* node.traverseBoth();
+    }
+  }
+
+  @override
+  Iterable<K> traverseKeys() sync* {
+    for (final node in _nodes) {
+      yield* node.traverseKeys();
+    }
+  }
+
+  @override
+  Iterable<V> traverseValues() sync* {
+    for (final node in _nodes) {
+      yield* node.traverseValues();
+    }
+  }
+}
+
+/// A [BitmapIndexNode] is used for compact representations
+///
+/// uses a bitmap to compact the list of nodes.
+class BitmapIndexNode<K, V> extends Node<K, V> {
+  final List<Node<K, V>> _nodes;
+  final int _shift;
+  final int _bitmap;
+
+  ///
+  BitmapIndexNode(this._nodes, this._shift, this._bitmap);
+
+  ///
+  BitmapIndexNode.empty(this._shift)
+      : _nodes = const [],
+        _bitmap = 0;
+
+  int _index(int bit) => popcount(_bitmap & (bit - 1));
+
+  @override
+  V find(int hash, K key) {
+    final bit = bitpos(hash, _shift);
+    if ((_bitmap & bit) != 0) {
+      return _nodes[_index(bit)].find(hash, key);
+    }
+    return null;
+  }
+
+  @override
+  Node<K,V> remove(int shift, int hash, K key) {
+    final bit = bitpos(hash, _shift);
+    if ((_bitmap & bit) != 0) {
+      // TODO: compaction to null if empty.
+      final newNodes = new List<Node<K,V>>.from(_nodes)..removeAt(_index(bit));
+      return new BitmapIndexNode(newNodes, _shift, _bitmap ^ bit);
+    } else {
+      return this;
+    }
+  }
+
+  @override
+  Node<K,V> assoc(int shift, int hash, K key, V value) {
+    int bit = bitpos(hash, shift);
+		int index = _index(bit);
+		if((_bitmap & bit) != 0) {
+			final newNodes = new List<Node<K,V>>.from(_nodes, growable: false);
+      newNodes[index] = _nodes[index].assoc(shift + 5, hash, key, value);
+      return new BitmapIndexNode(newNodes, shift, _bitmap);
+		} else {
+		    int n = popcount(_bitmap);
+        if (n >= 16) {
+          // convert to [ArrayNode]
+          final newNodes = new List<Node<K,V>>.filled(32, null, growable: false);
+          newNodes[mask(hash, shift)] = new LeafNode<K,V>(key, value);
+          int j = 0;
+          for (int i = 0; i < 32; i++) {
+            if (((_bitmap >> i) & 1) != 0) {
+              newNodes[i] = _nodes[j];
+              j++;
+            }
+          }
+          return new ArrayNode<K,V>(newNodes, shift, n);
+        } else {
+          final newNodes = new List<Node<K,V>>(n + 1);
+          for (int i = 0; i < n; i++) {
+            newNodes[i] = _nodes[i];
+          }
+          newNodes[index] = new LeafNode<K,V>(key, value);
+          return new BitmapIndexNode<K,V>(newNodes, shift, _bitmap | bit);
+        }
+		}
+  }
+
+  @override
```

Comment:
Since these three methods are the same in the array node and bitmap node classes, you might want to put their implementations in the abstract parent class.